compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -692,32 +673,28 @@ impl<'tcx> Cx<'tcx> {
                     };
 
                     let source = if let Some((did, offset, var_ty)) = var {
-                        let mk_const = |literal| Expr {
+                        let kind =
+                            ExprKind::ScalarLiteral { lit: (offset as u128).into(), user_ty: None };
+                        let offset = self.thir.exprs.push(Expr {
                             temp_lifetime,
-                            ty: var_ty,
+                            ty: expr_ty,
```

Comment:
This looks suspicious ^^ you're using the cast expression type here, which in your failing example is u64, but the literal is u32